### ISSUE TITLE
Gradle Kotlin DSL 0.14.0 shipped with Gradle 4.5 or later brakes binary plugin compatablity

### ISSUE THREAD

**Issue by sic-bre** [gradle/kotlin-dsl#719]
Starting with Gradle 4.5 / 4.5.1 our in house Kotlin DSL plugins are no longer binary compatible with older versions of Gradle (<= Gradle 4.4.1).  It appears to us that the change to Kotlin 1.2 in the engine that was introduced with the Gradle Kotlin DSL 0.14.0 is the root cause of this compatibility issue.  We get warnings listed below.

Note:  Forward compatibly of plugins compiled under Gradle 4.4.1 but used by Gradle 4.5.1 seems to work.

### Expected Behavior

I understand why the change from Kotlin 1.1 to 1.2 broke backwards compatibility.  However I found no documentation of this limitation in the release notes.  I feel that a binary compatibility change should be limited to significant version releases and be well documented.  We are now having to do significant work to decide if we are going to migrate all of our plugins AND client builds to >= Gradle 4.5, or if we are going to have to lock down our plugins to be only built in <= Gradle 4.4.1 for the foreseeable future.

Also if a plugin has binary incompatibles a clear error message should be printed at startup as to the root cause (Kotlin incompatibility) and prevent running rather than crashing or only partially working.

### Current Behavior

With plugins compiled under Gradle 4.5.1 we see the following error with Gradle 4.4.1:

$ /c/Program\ Files/Gradle/gradle-4.4.1/bin/gradle --no-daemon clean

FAILURE: Build failed with an exception.

* What went wrong:
org/gradle/kotlin/dsl/ProjectExtensionsKt$sam$Action$i$c0412805
> org.gradle.kotlin.dsl.ProjectExtensionsKt$sam$Action$i$c0412805

Note the same plugin compiled with Gradle 4.4.1 with no code changes does not have this issue.

### Context

All of our builds broke when we upgraded to plugins to use Gradle 4.5.1.  We had to back out of this change until EVERY build in our environment is using >= Gradle 4.5 as our Kotlin DSL plugins are used with every build job.

### Steps to Reproduce (for bugs)

Build any plugin that uses Kotlin 1.2 (>= Gradle 4.5) changed features such as Single Abstract Methods and try to use with Gradle 4.5.1

### Your Environment

Linux and Windows using Gradle 4.3.1, 4.4.1 and 4.5.1 with Kotlin DSL written in house plugins.

**Comment by eskatos** [comment]
Thanks for the comprehensive report @sic-bre.

Upgrade to Kotlin 1.2.0 is mentioned as a potential breaking change in the [release notes](https://docs.gradle.org/4.5.1/release-notes.html#kotlin-upgraded-to-1.2.0). Also see our [Kotlin versioning policy](https://github.com/gradle/kotlin-dsl/blob/master/doc/getting-started/Kotlin-Versions.md), especially this part:

> Until the release of Gradle Kotlin DSL v1.0 our policy will be to ship with the latest stable Kotlin version available at the time.

We are aware of the inconvenience and plan to handle this starting with Gradle Kotlin DSL v1.0. Please see #679.

**Comment by eskatos** [comment]
Closing as a duplicate of #679

**Comment by sic-bre** [comment]
Thanks for the feedback.  I must have missed #679 when I reviewed the list of existing open issues.

I did read the release notes, my comment was more targeted that the release notes only said "may break the build" but did not talk at all about binary compatibility between plugins.  My builds continued to work great as all my test cases at the time synchronized building the plugins and build scripts.  It was only later that I found issues with builds with outdated gradlew wrappers broke on the newly compiled plugins.

Hopefully this issue and #679 will give people a heads up on the potential compatibility issues.

I've been an early adopter of Kotlin DSL and love working with it.  I am looking forward to the stability of a v1.0 release should bring.
